### DISCUSSION TITLE
Why are there NaN in my PredictedPoint arrays?

### DISCUSSION THREAD

**Discussion by cxrodgers** [talmolab/sleap#1603] [Help!] [answered]
Hi all! Hopefully quick question. I'm using sleap to make predictions by interacting directly with my model from ipython, following this guide:
https://sleap.ai/develop/notebooks/Interactive_and_realtime_inference.html

Here's the syntax I'm using:
```
In [20]: predictions = predictor.predict(video[frame_numbers], make_labels=True)
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 388.9 FPS
```

What's a bit weird is that I'm getting some null (NaN) predictions. Specifically:
```
array([(394.34771729,  54.91504288,  True, False, 0.86732537),
       (499.03039551, 163.7522583 ,  True, False, 0.90653384),
       (495.5881958 , 319.40463257,  True, False, 0.87278712),
       (386.1685791 , 423.6703186 ,  True, False, 0.92618001),
       (234.28677368, 417.76049805,  True, False, 0.88510716),
       (129.32424927, 307.52090454,  True, False, 0.83302051),
       (135.56903076, 153.96514893,  True, False, 0.88759738),
       (246.1832428 ,  49.15915298,  True, False, 0.85349309),
       (336.46188354, 186.68826294,  True, False, 0.88542616),
       (368.32012939, 218.66969299,  True, False, 0.87827206),
       (366.58135986, 261.60397339,  True, False, 0.85256714),
       (336.70596313, 293.72000122,  True, False, 0.86500794),
       (293.61218262, 290.53265381,  True, False, 0.73720282),
       (262.09402466, 261.59783936,  True, False, 0.69810158),
       (260.38137817, 217.20942688,  True, False, 0.78168488),
       (293.72106934, 184.83102417,  True, False, 0.85220373),
       (         nan,          nan,  True, False, 0.        ),
       (404.89883423, 154.47032166,  True, False, 0.27019042),
       (438.89575195, 240.23535156,  True, False, 0.36890131),
       (401.56271362, 327.98712158,  True, False, 0.53179657),
       (311.22305298, 361.48291016,  True, False, 0.46658254),
       (         nan,          nan,  True, False, 0.        ),
       (         nan,          nan,  True, False, 0.        ),
       (229.85169983, 147.67066956,  True, False, 0.3094956 )],
      dtype=(sleap.instance.PredictedPoint, [('x', '<f8'), ('y', '<f8'), ('visible', '?'), ('complete', '?'), ('score', '<f8')])).view(numpy.recarray)
```

What I would like to do is see the prediction value, even if it's a totally garbage prediction with very low confidence. (For technical reasons I don't care about the accuracy of the predictions for this analysis.)

What's strange is that these predictions are null even when the `visible` column is True. So it doesn't have to do with the logic inside LabeledFrame.numpy() that masks out invisible nodes. I do notice that these null predictions have a `score` of 0. I tried setting `predictor.peak_threshold` to zero and then to a negative number but it didn't change anything. 

I wonder if this is because I recently added some nodes to my skeleton after the initial training set, meaning many of the training frames have missing data for some nodes. These are indeed the nodes that are most likely to show up as null. However, the same nodes can be predicted on other frames, it's not like they can never be predicted.

Thanks for any tips!

**Comment by talmo** [comment, marked as answer]
Hi @cxrodgers,

You can get this by adjusting the score threshold when you load up the model:

```py
predictor = sleap.load_model(model_path, peak_threshold=1e-3)
```

You can set it to 0 or a negative number, but you'll get a LOT of false positives. At a very low threshold you'll get almost all non-NaNs, even if they're mostly noise. Anything below a very low value (~1e-3) is just going to give you peaks due to numerical rounding errors.

Give that a go and let us know if that works for your use case!

Cheers,

Talmo

**Reply by cxrodgers** [reply]
Thanks, this works!

I think my model is confused and under-confident because the nodes I added to the skeleton later on are marked as "invisible" in the original training set, even though they're clearly visible. I'm messing around with changing the confidence threshold post-hoc, which is why I need to extract even the noisy points. 

I got pretty close to figuring this out (`predictor.peak_threshold = 1e-3`) but didn't realize it needed to be in the `sleap.load_model` instead. I'm happy to make a doc PR if you want it but I don't see `sleap.load_model` documented anywhere. Hopefully at least this question will be a resource for anyone with the same problem.
